File: tools/parameter_changes_layout.zig
const std = @import("std");
const parameter_changes = @import("zig-vst3").pluginterfaces.vst.ivstparameterchanges;

pub fn main() !void {
    const stdout = std.io.getStdOut().writer();
    try printType(stdout, "IParamValueQueue", parameter_changes.IParamValueQueue);
    try printType(stdout, "IParameterChanges", parameter_changes.IParameterChanges);

    try printTuid(stdout, "IParamValueQueue", parameter_changes.iparam_value_queue_iid);
    try printTuid(stdout, "IParameterChanges", parameter_changes.iparameter_changes_iid);
}

fn printType(writer: anytype, comptime name: []const u8, comptime Type: type) !void {
    try writer.print("{s} size {} align {}\n", .{ name, @sizeOf(Type), @alignOf(Type) });
}

fn printTuid(writer: anytype, comptime name: []const u8, bytes: [16]u8) !void {
    try writer.print("{s} iid", .{name});
    for (bytes) |byte| {
        try writer.print(" {X:0>2}", .{byte});
    }
    try writer.writeByte('\n');
}

File: tests/abi/parameter_changes_layout.cpp
#include "pluginterfaces/vst/ivstparameterchanges.h"

#include <cstddef>
#include <cstdio>

#define PRINT_TYPE(Type) std::printf (#Type " size %zu align %zu\n", sizeof (Steinberg::Vst::Type), alignof (Steinberg::Vst::Type))

template <typename T>
static void print_iid (const char* name, const T& tuid)
{
	std::printf ("%s iid", name);
	for (int index = 0; index < 16; ++index)
		std::printf (" %02X", static_cast<unsigned char> (tuid[index]));
	std::printf ("\n");
}

int main ()
{
	PRINT_TYPE (IParamValueQueue);
	PRINT_TYPE (IParameterChanges);

	print_iid ("IParamValueQueue", Steinberg::Vst::IParamValueQueue_iid);
	print_iid ("IParameterChanges", Steinberg::Vst::IParameterChanges_iid);
	return 0;
}
